10장_예외/item71.md

```diff
@@ -0,0 +1,198 @@
+  * 비검사 예외: 비검사 예외(UnChecked Exception)는 Runtime Exception를 상속한 예외로 **프로그래머의 실수로 발생하는 예외들입니다. 비검사 예외는 검사 예외와 달리 예외 처리를 강제하지 않습니다.** 런타임 에러를 발생시키는 에러들로 처리할수 없기 때문에 예외 처리를 강제하지 않습니다. 다만, 비검사예외도 예외 처리를 할 수 있기 때문에 예외 처리만 한다면 런타임 에러는 발생하지 않습니다, 하지만 좋지 않은 방법입니다. 그로 인한 더 큰 sideEffect가 일어날 겁니다.(알고리즘 문제를 `IndexOutOfBoundsException`을 예외처리하여 성공시킨 코드를 한번쯤은 보셨을 겁니다). 대표적인 예외로 `IndexOutOfBoundsException`, `NullPointerException` 등이 있습니다. 
+
+## 예외처리를 해야하는 검사 예외, 잘 쓰면 약 못 쓰면 독.
+
+검사 예외를 제대로 활용하면 API와 프로그램 질을 높일 수 있습니다. 결과를 코드로 반환하거나 비검사 예외를 던지는 것과 달리, **검사 예외는 발생한 문제를 프로그래머가 처리**하여 안정성을 높이게끔 해줍니다. 
+
+하지만 검사 예외를 과도하게 사용하면 오히려 불편한 API가 됩니다. 메서드가 검사 예외를 던질 수 있다고 선언됐다면, 이를 호출하는 코드에서는 catch 블록을 두어 **그 예외를 붙잡아 처리하거나 더 바깥으로 던져 문제를 전파**해야만 합니다.
+
+더구나 검사 예외를 던지는 메서드는 **스트림 안에서 직접 사용할 수 없기 때문에** 자바 8부터는 부담이 더욱 커집니다.
+
+따라서 API를 제대로 사용해도 발생할 수 있는 예외이거나, 프로그래머가 의미 있는 조치를 취할 수 있는 경우라면 이 정도 부담쯤은 받아들일 수 있을 것입니다. 
+
+### 그러나 둘 중 어디에도 해당하지 않는다면 비검사 예외를 사용하는게 좋습니다. 
+
+* 참조한 곳: https://pridiot.tistory.com/54, https://reference-m1.tistory.com/246
+
+<br>
+
+# Objective-C
+
+* Objective-C 에서는 에러가 크게 NSError와 NSException이 있습니다. 
+  * NSError(Class Type): NSError는 치명적이지 않은, 개발자가 처리 가능한 에러입니다. NSError에 의해 캡처되도록 설계된 문제는 종종 사용자 오류 (또는 사용자에게 표시 될 수있는 오류)입니다. 
+  (따라서 -presentError : 및 NSErrorRecoveryAttempting)에서 종종 복구 될 수 있습니다.
+  일반적으로 예상되거나 예측 가능한 오류입니다(예 : 액세스 할 수없는 파일을 열려고 시도하는 경우,
+  또는 호환되지 않는 문자열 인코딩 간 변환 시도).
+
+  * NSException: NSException은 잠재적으로 치명적인 **프로그래머 오류**를 위해 설계되었습니다. 이러한 오류는 일부 작업을 수행하기위한 사전 조건을 올바르게 확인하지 않은 애플리케이션의 잠재적 인 결함을 나타 내기 위해 설계되었습니다 (예 : 경계를 벗어난 배열 인덱스에 액세스하려고 시도하거나 변경 불가능한 객체를 변경하려고 시도 함).
+
+
+```objective-c
+NSException *exception = [[NSException alloc] initWithName:NSRangeException reason:@"" userInfo:nil];
+```
+=> Objective-c 에서는 NSRangeException이 에러 이름(스트링 타입)으로서 존재합니다.
+
+### NSException도 자바의 비검사 예외와 마찬가지로 @try - @catch 문으로 해결할 수 있습니다. 하지만 프로그래머의 실수를 이렇게 해결하는 건 좋지 않은 방법입니다.  
+
+```objective-c
+@try {
+    NSArray *array = [[NSArray alloc] initWithObjects:@"aaa", nil];
+    NSString *result = [array objectAtIndex:5];
+} @catch (NSException *exception) {
+    NSLog(@"%@ is exception name", exception.name);
+    NSLog(@"%@ is exception reason", exception.reason);
+      
+} @finally {
+    NSLog(@"finally");
+}
+```
+
+참고한 곳: https://stackoverflow.com/questions/11100951/nsexception-and-nserror-custom-exception-error, https://developer.apple.com/library/archive/documentation/Cocoa/Conceptual/ObjectiveC/Introduction/introObjectiveC.html#//apple_ref/doc/uid/TP30001163
+
+
+# Swift
+## Swift is made to be typeSafe language: 런타임 에러는 처리가 아닌 예방으로
+
+* Swift의 에러는 크게 Error 와 Runtime Exception으로 나눌 수 있겠습니다. 
+  * Error(enumType): 처리 가능한 에러들로 **무조건 에러 처리를 해야 합니다** 에러를 throw 하거나 catch 해서 에러를 잡아냅니다. 처리 가능한 에러 이므로 main 함수까지 계속 던지지만 않는다면 런타임에러는 발생하지 않습니다. 
+  * Runtime Error: 런타임 예외 (trap, atal error, assertion failure, etc 등이라고도 함)는 프로그래머가 발생시키는 오류입니다. 
+  -Ounchecked 빌드를 제외하고 Swift는 **일반적으로 불량 / 정의되지 않은 상태에서 계속 실행하는 대신 프로그램이 크래시가 발생할 것입니다.**
```

Comment:
내용에 관한건 아니지만 `Ounchecked 빌드`에 대해서 궁금해서 찾아봤는데요

[Zedd님의 2018년 자료](https://zeddios.tistory.com/593)에는 
> - Ounchecked : 안전성(safety)를 성능으로 교환할 의향이 있는 특정 라이브러리, 앱을 위한 특수한 최적화 모드입니다. 컴파일러는 모든 overflow검사와 일부 implicit(암시적) 타입 검사를 제거합니다. 하지만 이는 감지되지 않은 메모리 문제와 정수 overflow를 초래할 수 있기 때문에 일반적으로 사용되지 않습니다. 정수 overflow 및 타입 캐스트와 관련하여 코드가 안전하다는 것을 신중하게 검토 경우에만 이 Ounchecked를 사용합니다.
출처: https://zeddios.tistory.com/593 [ZeddiOS]

라고 나와있는데 현재에는 없는 것 같네요!

[Github Apple Swift Docs 2021.05](https://github.com/apple/swift/blob/main/docs/OptimizationTips.rst#enabling-optimizations)

> -Onone: This is meant for normal development. It performs minimal optimizations and preserves all debug info.
-O: This is meant for most production code. The compiler performs aggressive optimizations that can drastically change the type and amount of emitted code. Debug information will be emitted but will be lossy.
-Osize: This is a special optimization mode where the compiler prioritizes code size over performance.

[Crash early in Swift - 2019](https://uynguyen.github.io/2019/01/19/Crash-early-in-Swift/)에서
> Updates: As you can see **there is no longer the -Ounchecked mode** in Xcode10, instead a new option introduced Optimize for Size. The main difference between the O mode and Osize mode is “When compiling with -O the compiler tries to transform the code so that it executes with maximum performance. However, this improvement in runtime performance can sometimes come with a tradeoff of increased code size. With the new -Osize optimization mode the user has the choice to compile for minimal code size rather than for maximum speed” (swift.org)

수정할 건 없지만 참고하면 좋을 것 같아 공유합니다!